File: 2022/day_02/c/part2.c
/**
 * @author: DevHeadTech
 * 
 * Advent of Code - Day 02
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <errno.h>
#include <math.h>

#define LOSE    0
#define TIE     3
#define WIN     6

#define NORM_MOVE      64
#define NORM_ROUND     89

bool open_file(const char *path, FILE** file)
{
    errno = 0;
    *file = fopen(path,"r");

    if (NULL == *file || errno != 0 )
    {
        perror("fopen");
        return false;
    }

    return true;
}

void process_file(FILE* file)
{
    int ret = 0;
    int opp_move = 0;
    int my_move = 0;
    char buf[16] = {0};
    uint opp_score = 0;
    uint my_score = 0;
    int outcome = 0;
    // Scissors, Rock, Paper
    int move[] = {3,1,2};

    while (NULL != fgets(buf, 16, file))
    {   // scanf doesn't clear the whole int
        opp_move = 0;
        outcome = 0;
        if (EOF != sscanf(buf, "%c %c", &opp_move, &outcome))
        {   // normalize
            outcome -= NORM_ROUND;
            opp_move -= NORM_MOVE;
            opp_score += opp_move;
            my_move = move[(opp_move+outcome)%3];
            my_score += my_move;

            if (0 == outcome)
            {   // tie
                my_score += 3;
                opp_score += 3;
            }
            else if(-1 == outcome)
            {   // lose
                opp_score += 6;
            }
            else
            {   // win
                my_score += 6;
            }

            //printf("opp: %d my: %d\n", opp_move, my_move);
        }
    } 

    printf("My Score: %d\n", my_score);
    printf("Opponent Score: %d\n", opp_score);

}

int main(int argc, char** argv)
{
    FILE* file = NULL;

    // only one input allowed
    if (argc < 2 || argc > 2)
    {
        fprintf(stderr, "Bad Parameters\n");
        // todo: print usage
        return EXIT_FAILURE;
    }

    if (open_file(argv[1], &file))
    {
        process_file(file);
        fclose(file);
        return EXIT_SUCCESS;
    }

    return EXIT_FAILURE;
}

File: 2022/day_02/python/part2.py
#
# @author: DevHeadTech
# 
# Advent of Code - Day 02
#

import sys

opp_dict = {
    "A":1,
    "B":2,
    "C":3
}

outcome_dict = {
    "X":-1,
    "Y":0,
    "Z":1
}

actions = [3,1,2]

if len(sys.argv) < 2 or len(sys.argv) > 2:
    print('Bad Parameters')
#

def process(round):
    global my_score 

    outcome = outcome_dict[round[1]]
    opp_action = opp_dict[round[0]]
    my_action = actions[(opp_action+outcome)%3]
    my_score += my_action

    if(outcome == 0): #tie
        my_score += 3
    elif(outcome == 1): #win
        my_score += 6
#

file = open(sys.argv[1], "r")

my_score = 0 
while ( line := file.readline() ):
    if '' != line.strip():
        round = line.split()
        process(round)
    #
#

print("My Score: ", my_score)

file.close()
